src/main/java/net/dries007/tfc/common/blockentities/BarrelBlockEntity.java

```diff
@@ -0,0 +1,478 @@
+    public void setLastUpdateTick(long tick)
+    {
+        lastUpdateTick = tick;
+    }
+
+    @Override
+    public void ejectInventory()
+    {
+        super.ejectInventory();
+        assert level != null;
+        inventory.excess.stream().filter(item -> !item.isEmpty()).forEach(item -> Helpers.spawnItem(level, worldPosition, item));
+    }
+
+    public void onSeal()
+    {
+        assert level != null;
+        if (!level.isClientSide())
+        {
+            // Drop container items, but allow the main slot to be filled
+            for (int slot : new int[] { SLOT_FLUID_CONTAINER_IN, SLOT_FLUID_CONTAINER_OUT })
+            {
+                Helpers.spawnItem(level, worldPosition, inventory.getStackInSlot(slot));
+                inventory.setStackInSlot(slot, ItemStack.EMPTY);
+            }
+        }
+
+        sealedTick = Calendars.get(level).getTicks();
+        updateRecipe();
+        if (recipe != null)
+        {
+            recipe.onSealed(inventory);
+            recipeTick = sealedTick;
+        }
+        markForSync();
+    }
+
+    public void onUnseal()
+    {
+        sealedTick = recipeTick = 0;
+        if (recipe != null)
+        {
+            recipe.onUnsealed(inventory);
+        }
+        updateRecipe();
+        markForSync();
+    }
+
+    protected void updateRecipe()
+    {
+        assert level != null;
+
+        final SealedBarrelRecipe oldRecipe = recipe;
+        if (inventory.excess.isEmpty())
+        {
+            // Will only work on a recipe as long as the 'excess' is empty
+            recipe = level.getRecipeManager().getRecipeFor(TFCRecipeTypes.BARREL_SEALED.get(), inventory, level).orElse(null);
+            if (recipe != null && oldRecipe != recipe)
+            {
+                // The recipe has changed to a new one, so update the recipe ticks
+                recipeTick = Calendars.get(level).getTicks();
+                markForSync();
+            }
+        }
+
+        needsRecipeUpdate = false;
+    }
+
+    private void updateFluidIOSlots()
+    {
+        assert level != null;
+        ItemStack input = inventory.getStackInSlot(SLOT_FLUID_CONTAINER_IN);
+        if (!input.isEmpty() && inventory.getStackInSlot(SLOT_FLUID_CONTAINER_OUT).isEmpty())
+        {
+            input.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY).ifPresent(cap -> {
+                if (cap.getFluidInTank(0).getAmount() > 0)
+                {
+                    final int maxFill = TFCConfig.SERVER.barrelCapacity.get() - inventory.tank.getFluid().getAmount();
+                    if (maxFill > 0)
+                    {
+                        if (inventory.tank.fill(cap.drain(maxFill, IFluidHandler.FluidAction.SIMULATE), IFluidHandler.FluidAction.SIMULATE) > 0)
+                        {
+                            if (inventory.tank.fill(cap.drain(maxFill, IFluidHandler.FluidAction.EXECUTE), IFluidHandler.FluidAction.EXECUTE) > 0)
+                            {
```

Comment:
This sequence should probably use `FluidHelpers.transferUpTo(, ... , TFCConfig.SERVER.barrelCapacity.get())`. It does this check a bit more robustly.